src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -28,226 +20,168 @@ import io.iohk.ethereum.ledger.BlockData
     * @param blockExecutionScheduler threadPool on which the execution should be run
     * @param blockchainConfig        blockchain configuration
     * @return One of:
-    *   - [[BlockImportedToTop]] - if the block was added as the new best block
-    *   - [[BlockEnqueued]]      - block is stored in the [[io.iohk.ethereum.ledger.BlockQueue]]
-    *   - [[ChainReorganised]]   - a better new branch was found causing chain reorganisation
-    *   - [[DuplicateBlock]]     - block already exists either in the main chain or in the queue
-    *   - [[BlockImportFailed]]  - block failed to execute (when importing to top or reorganising the chain)
+    *   - [[ExtendedCurrentBestBranch]] - if the branch was added on top of the current branch
+    *   - [[SelectedNewBestBranch]]     - if the chain was reorganized.
+    *   - [[KeptCurrentBestBranch]]     - if the branch was not considered as better than the current branch
+    *   - [[ConsensusError]]            - block failed to execute (when importing to top or reorganising the chain)
+    *   - [[ConsensusErrorDueToMissingNode]]  - block failed to execute (when importing to top or reorganising the chain)
     */
   override def evaluateBranch(
-      branch: Seq[Block]
-  )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[BlockImportResult] = {
-    val block = branch.head
+      branch: NonEmptyList[Block]
+  )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[ConsensusResult] =
     blockchainReader.getBestBlock() match {
       case Some(bestBlock) =>
         blockchain.getChainWeightByHash(bestBlock.header.hash) match {
-          case Some(weight) => handleBlockImport(block, bestBlock, weight)
+          case Some(weight) => handleBranchImport(branch, bestBlock, weight)
           case None         => returnNoTotalDifficulty(bestBlock)
         }
       case None => returnNoBestBlock()
     }
-  }
 
-  private def handleBlockImport(block: Block, bestBlock: Block, weight: ChainWeight)(implicit
+  private def handleBranchImport(
+      branch: NonEmptyList[Block],
+      currentBestBlock: Block,
+      currentBestBlockWeight: ChainWeight
+  )(implicit
       blockExecutionScheduler: Scheduler,
       blockchainConfig: BlockchainConfig
-  ): Task[BlockImportResult] = {
-    val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
-      importToTop(block, bestBlock, weight)
-    } else {
-      reorganiseOrEnqueue(block, bestBlock, weight)
+  ): Task[ConsensusResult] = {
+    val parentHash = branch.head.header.parentHash
+    val parentWeight = blockchain
+      .getChainWeightByHash(parentHash)
+      .getOrElse(throw new Error("Inconsistent database"))
+
+    val consensusResult: Task[ConsensusResult] =
+      if (newBranchWeight(branch, parentWeight) > currentBestBlockWeight) {
+        if (currentBestBlock.isParentOf(branch.head)) {
```

Comment:
Wouldn't it make more sense to do this tests first (`currentBestBlock.isParentOf(branch.head)`) and only after calculate the `newBranchWeight` if necessary?

Reply:
Also we could save a read on getting `parentWeight`, and only do it if needed

Reply:
Yes. I note however that in the final design we would not really care about the fact that we are extending the branch or changing the branch, so this condition is there for compatibility but I think in the end it might just be the weight comparison.